### PR TITLE
CLI: Fix kontena grid cloud-config network units

### PR DESCRIPTION
**Pull request by SpComb**
Fix `kontena grid cloud-config` for current #1264 CoreOS stable systemd-networkd issues.

This should match the cloud-config that the plugins are using for provisioning, and the `zz-default.network` `[Match] Name=eth*` matches what this command has always generated.

OTOH, the previous `00-eth.network` file matching `eth*` would NOT override the `zz-default.network`, so it is possible that this updated cloud-config would break networking, if the previous `00-eth.network` was not used, and this replaces the previously used `zz-default.network`.

## Usage

```
$ kontena grid cloud-config test > cloud-config.yaml
 [warn] Guessing --default-interface-match=eth* from --peer-interface=eth1, make sure that this matches the interface names used by the node platform
$ kontena grid cloud-config --peer-interface=em5p1 test > cloud-config.yaml
 [error] Unable to determine --default-interface-match from --peer-interface=em5p1, configure --default-interface-match= explicitly
$ kontena grid cloud-config --peer-interface=em5p1 --default-interface-match=em*p* test > cloud-config.yaml 
```

## Testing

Seems to work as intended for an existing node

```
core@hidden-river-52 ~ $ sudo coreos-cloudinit -from-file cloud-config.yaml
2017/01/04 16:12:26 Checking availability of "local-file"
2017/01/04 16:12:26 Fetching user-data from datasource of type "local-file"
2017/01/04 16:12:26 Fetching meta-data from datasource of type "local-file"
2017/01/04 16:12:26 Parsing user-data as cloud-config
2017/01/04 16:12:26 Merging cloud-config from meta-data and user-data
2017/01/04 16:12:26 Writing file to "/etc/kontena-agent.env"
2017/01/04 16:12:26 Wrote file to "/etc/kontena-agent.env"
2017/01/04 16:12:26 Wrote file /etc/kontena-agent.env to filesystem
2017/01/04 16:12:26 Writing file to "/etc/systemd/system/docker.service.d/50-kontena.conf"
2017/01/04 16:12:26 Wrote file to "/etc/systemd/system/docker.service.d/50-kontena.conf"
2017/01/04 16:12:26 Wrote file /etc/systemd/system/docker.service.d/50-kontena.conf to filesystem
2017/01/04 16:12:26 Writing file to "/etc/sysctl.d/99-inotify.conf"
2017/01/04 16:12:26 Wrote file to "/etc/sysctl.d/99-inotify.conf"
2017/01/04 16:12:26 Wrote file /etc/sysctl.d/99-inotify.conf to filesystem
2017/01/04 16:12:26 Updated /etc/environment
2017/01/04 16:12:26 Masking unit file "10-weave.network"
2017/01/04 16:12:26 Masking unit file "50-docker.network"
2017/01/04 16:12:26 Masking unit file "50-docker-veth.network"
2017/01/04 16:12:26 Writing unit "zz-default.network" to filesystem
2017/01/04 16:12:26 Writing file to "/etc/systemd/network/zz-default.network"
2017/01/04 16:12:26 Wrote file to "/etc/systemd/network/zz-default.network"
2017/01/04 16:12:26 Wrote unit "zz-default.network"
2017/01/04 16:12:26 Writing unit "kontena-agent.service" to filesystem
2017/01/04 16:12:26 Writing file to "/etc/systemd/system/kontena-agent.service"
2017/01/04 16:12:26 Wrote file to "/etc/systemd/system/kontena-agent.service"
2017/01/04 16:12:26 Wrote unit "kontena-agent.service"
2017/01/04 16:12:26 Enabling unit file "kontena-agent.service"
2017/01/04 16:12:26 Enabled unit "kontena-agent.service"
2017/01/04 16:12:26 Ensuring runtime unit file "etcd.service" is unmasked
2017/01/04 16:12:26 Ensuring runtime unit file "etcd2.service" is unmasked
2017/01/04 16:12:26 Ensuring runtime unit file "fleet.service" is unmasked
2017/01/04 16:12:26 Ensuring runtime unit file "locksmithd.service" is unmasked
2017/01/04 16:12:26 Restarting systemd-networkd
2017/01/04 16:12:27 Restarted systemd-networkd (done)
2017/01/04 16:12:27 Calling unit command "start" on "kontena-agent.service"'
2017/01/04 16:12:27 Result of "start" on "kontena-agent.service": done
core@hidden-river-52 ~ $ networkctl 
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           carrier     unmanaged 
  2 eth0             ether              routable    configuring
  3 eth1             ether              routable    configuring
  4 docker0          ether              no-carrier  unmanaged 
  5 datapath         ether              degraded    unmanaged 
  7 weave            ether              routable    unmanaged 
  8 dummy0           ether              off         unmanaged 
 10 vethwe-datapath  ether              degraded    unmanaged 
 11 vethwe-bridge    ether              degraded    unmanaged 
 12 vxlan-6784       ether              degraded    unmanaged 

10 links listed.
core@hidden-river-52 ~ $ ls -laR /{etc,lib,run}/systemd/network/
/etc/systemd/network/:
total 36
drwxr-xr-x. 2 root root 4096 Jan  4 16:12 .
drwxr-xr-x. 6 root root 4096 Jan  3 12:08 ..
lrwxrwxrwx. 1 root root    9 Jan  4 16:12 10-weave.network -> /dev/null
lrwxrwxrwx. 1 root root    9 Jan  4 16:12 50-docker-veth.network -> /dev/null
lrwxrwxrwx. 1 root root    9 Jan  4 16:12 50-docker.network -> /dev/null
-rw-r--r--. 1 root root  213 Jan  4 16:12 zz-default.network

/lib/systemd/network/:
total 104
drwxr-xr-x.  2 root root 4096 Dec  7 09:53 .
drwxr-xr-x. 13 root root 4096 Dec  7 09:51 ..
-rw-r--r--.  1 root root   31 Dec  7 09:31 50-docker-veth.network
-rw-r--r--.  1 root root   44 Dec  7 09:31 50-docker.network
-rw-r--r--.  1 root root   33 Dec  7 09:21 50-flannel.network
-rw-r--r--.  1 root root  605 Dec  7 09:28 80-container-host0.network
-rw-r--r--.  1 root root  678 Dec  7 09:28 80-container-ve.network
-rw-r--r--.  1 root root  664 Dec  7 09:28 80-container-vz.network
-rw-r--r--.  1 root root  277 Dec  7 09:31 98-virtio.link
-rw-r--r--.  1 root root   80 Dec  7 09:28 99-default.link
-rw-r--r--.  1 root root  109 Dec  7 09:31 yy-pxe.network
-rw-r--r--.  1 root root  108 Dec  7 09:31 yy-vmware.network
-rw-r--r--.  1 root root   55 Dec  7 09:31 zz-default.network

/run/systemd/network/:
total 16
drwxr-xr-x.  2 root root 140 Jan  4 16:11 .
drwxr-xr-x. 18 root root 460 Jan  4 16:12 ..
-rw-r--r--.  1 root root 165 Jan  3 12:08 00-56:b4:c1:6f:d7:54.network
-rw-r--r--.  1 root root 299 Jan  3 12:08 00-86:57:db:26:05:99.network
-rw-r--r--.  1 root root 120 Jan  4 16:11 00-eth.network
-rw-r--r--.  1 root root  43 Jan  3 12:08 99-public-interface.network
lrwxrwxrwx.  1 root root  22 Jan  3 12:08 resolv.conf -> ../resolve/resolv.conf
```

### CODE REVIEW

File: cli/lib/kontena/machine/cloud_config/cloudinit.yml [hunk 1]
```diff
@@ -19,27 +19,28 @@ write_files:
...
-      runtime: true
+    - name: 10-weave.network
+      mask: true
+    - name: 50-docker.network
+      mask: true
+    - name: 50-docker-veth.network
+      mask: true
+    - name: zz-default.network
       content: |
+        # default should not match virtual Docker/weave bridge/veth network interfaces
```
Comment (jakolehm) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: Should we add warning that Match should be actually matching correct interfaces?

File: cli/lib/kontena/machine/cloud_config/cloudinit.yml [hunk 2]
```diff
@@ -19,27 +19,28 @@ write_files:
       fs.inotify.max_user_instances = 8192
 coreos:
   units:
-    - name: 00-eth.network
-      runtime: true
+    - name: 10-weave.network
```
Comment (jakolehm) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: Do we need this?

Comment (SpComb) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: I added this mainly to mask out any existing `10-weave.network` if upgrading the cloud-config for an existing node. If you have such a network unit, then you should remove it.

File: cli/lib/kontena/machine/cloud_config/cloudinit.yml (hunk 1 above)
Comment (SpComb) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: We could also try the `Path=*` trick, but I dunno exactly.

Maybe something related to the already existing `--peer-interface`?
```ruby
match_interface = peer_interface.sub(/\d+/, '*')
```

Comment (jakolehm) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: `--peer-interface` makes sense 👍

File: cli/lib/kontena/machine/cloud_config/cloudinit.yml (hunk 2 above)
Comment (SpComb) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: Removed the `10-weave.network` unit from the cloud-config.

I suppose it's not worth it to consider the use-case of re-generating the cloud-config for an existing node that was installed with an older cloud-config - you should just replace the entire node.

File: cli/lib/kontena/machine/cloud_config/cloudinit.yml (hunk 1 above)
Comment (SpComb) [inline cli/lib/kontena/machine/cloud_config/cloudinit.yml]: Added a `--default-interface-match` that defaults to the `--peer-interface=eth1` prefix, if it's of the form `abcN`.

**Comment by SpComb** [comment]
Tested to work with a hand-created DO droplet.

I think the way we configure the systemd-networkd DNS resolver/domain settings on the node is very fragile and prone to breakage, but that's a separate issue.

**Pull request merged**